Include deletions in modified files mask

jakeleveroni opened the pull request:
**Problem:**

The `git.getModifiedFiles` method does not include removed files.

**Solution:**

The diff filter being used does not include the deletions (D) mask. Changing the diff-filter from `ACMR` to `ACMRD` includes deleted files in the results of the `git.getModifiedFiles` method.

**Related issues:**

Fixes #706

**Checklist:**

- [ ] Added or updated tests
  - unsure of how to test this, suggestions are appreciated!
- [x] Added or updated documentation
- [x] Ensured the pre-commit hooks ran successfully

_By opening this pull request, you agree that this submission can be released under the same [License](https://github.com/paularmstrong/onerepo/blob/main/LICENSE.md) that covers the project._

paularmstrong reviewed (changes requested):
As implemented, this will break any command that relies on [`getFilepaths`](https://onerepo.tools/api/namespaces/builders/#getfilepaths) and passes them through for operating on. 

This is easy to see with eslint. In the oneRepo repository:
1. Delete `prettier.config.cjs`
2. Run `one lint`

```
one lint
 ┌ Filtering ignored files
 └ ✔ 43ms
 ┌ Lint files
 │ ERR Oops! Something went wrong! :(
 │ ERR
 │ ERR ESLint: 8.57.0
 │ ERR
 │ ERR No files matching the pattern "prettier.config.cjs" were found.
 │ ERR Please check for typing mistakes in the pattern.
 └ ✘ 1178ms
 ■ ✘ Completed with errors 1242ms
```

I don't think it's a good idea to change how `getFilepaths` works, as that would result in a breaking change here. So, in order to preserve that behavior, we should figure out which is best between:

1. Add a flag to `getModifiedFiles` to include deleted or not, eg `getModifiedFiles({ includeDeleted: true })`
2. OR – create a separate `getModifiedFilesByStatus` that returns an object of the shape:
   ```ts
   type ModifiedByStatus = {
     added: Array<string>;
     copied: Array<string>;
     modified: Array<string>;
     deleted: Array<string>;
     copied: Array<string>;
     updated: Array<string>;
   }
   ```
   
   This is likely a better long-term solution that would open more possibilities for other plugins and integrations, but may be slightly more work.

jakeleveroni commented:
> As implemented, this will break any command that relies on [`getFilepaths`](https://onerepo.tools/api/namespaces/builders/#getfilepaths) and passes them through for operating on.
> 
> This is easy to see with eslint. In the oneRepo repository:
> 
> 1. Delete `prettier.config.cjs`
> 2. Run `one lint`
> 
> ```
> one lint
>  ┌ Filtering ignored files
>  └ ✔ 43ms
>  ┌ Lint files
>  │ ERR Oops! Something went wrong! :(
>  │ ERR
>  │ ERR ESLint: 8.57.0
>  │ ERR
>  │ ERR No files matching the pattern "prettier.config.cjs" were found.
>  │ ERR Please check for typing mistakes in the pattern.
>  └ ✘ 1178ms
>  ■ ✘ Completed with errors 1242ms
> ```
> 
> I don't think it's a good idea to change how `getFilepaths` works, as that would result in a breaking change here. So, in order to preserve that behavior, we should figure out which is best between:
> 
> 1. Add a flag to `getModifiedFiles` to include deleted or not, eg `getModifiedFiles({ includeDeleted: true })`
> 2. OR – create a separate `getModifiedFilesByStatus` that returns an object of the shape:
>    ```ts
>    type ModifiedByStatus = {
>      added: Array<string>;
>      copied: Array<string>;
>      modified: Array<string>;
>      deleted: Array<string>;
>      copied: Array<string>;
>      updated: Array<string>;
>    }
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    This is likely a better long-term solution that would open more possibilities for other plugins and integrations, but may be slightly more work.

Thanks for the info, I also like the second option you mentioned. Ill try to get this updated this week or next week when i have some time. thanks for the feedback!

jakeleveroni commented:
@paularmstrong I updated the original getModifiedFiles method to take an additional parameter to specify whether or not to include deletions. But i also implemented the getModifiedFileByStatus method cause it seemed like it might be useful at some point. Not sure if you want to include it now or not but let me know, happy to pull it out if its not needed.

paularmstrong commented on modules/git/src/index.ts:
```diff
@@ -175,7 +184,11 @@ export type ModifiedStaged = {
-export async function getModifiedFiles(modified: ModifiedStaged | ModifiedFromThrough = {}, options: Options = {}) {
+export async function getModifiedFiles(
+	modified: ModifiedStaged | ModifiedFromThrough = {},
+	options: Options = {},
+	includeDeletions = false,
```
To follow general convention, let's merge `includeDeletions` into the `options: Options`. This keeps all of the methods from ballooning with too many unnamed arguments.

```ts
export type ModifiedFilesOptions = Options & { 
  /**
   * By default, this method will exclude files that were deleted to avoid accidentally attempting to operate on something that doesn't exist. To include the files in the output list, set this option to `true`.
   * @default `false`
   */
  includeDeletions?: boolean;
};

export async function getModifiedFiles(
  modified: Modified | ModifiedFromThrough = {},
  options: ModifiedFilesOptions = {};
) {
	const { from, staged, through } = modified;
	const { includeDeletions = false, step } = options;
// ...
```

paularmstrong reviewed (commented):
Nice. Almost there. I'm happy to take over on this if you don't have time.

paularmstrong commented:
> Additionally im not really sure how to write good tests around this since it requires git state mgmt, any pointers on how i can write meaningful tests for this?



> @paularmstrong I updated the original getModifiedFiles method to take an additional parameter to specify whether or not to include deletions. But i also implemented the getModifiedFileByStatus method cause it seemed like it might be useful at some point. I updated the tasks command to use the new method but it could just use the existing getModifiedFiles and pass true to the includeDeletions param. Not sure which way you prefer so lmk.

I think what you've done here is good!

> Additionally im not really sure how to write good tests around this since it requires git state mgmt, any pointers on how i can write meaningful tests for this?

I don't think I would write tests… there's a lot more setup necessary that's not really worth the payoff.

jakeleveroni commented:
> Nice. Almost there. I'm happy to take over on this if you don't have time.

Thanks! I will have time tomorrow to wrap up those changes.

paularmstrong commented:
@jakeleveroni I like what you did – and your changes helped me realize it could be simplified down to reuse the `getModifiedFiles` function just by including a single option. Take a look at the commit and let me know what you think.

jakeleveroni commented:
> @jakeleveroni I like what you did – and your changes helped me realize it could be simplified down to reuse the `getModifiedFiles` function just by including a single option. Take a look at the commit and let me know what you think.

Very nice, thats a clever implementation. I like it thanks for the assist!

paularmstrong commented:
> does that mean you would need to call the method like so to get the modified by status response: getModifiedFiles<true>({ byStatus: true })?

Nope! It infers it internally. The `getters` for `getFilepaths`, for example, automatically knows the return type is an array of strings because `byStatus` is not set / false.

I need to dig a little bit more to figure out why I had to drop in the `// @ts-expect-error` though, so not _quite_ done.

paularmstrong commented:
@jakeleveroni Sorry for the massive delay. I think the change in https://github.com/paularmstrong/onerepo/pull/718/commits/4ca2c65788077dc26bdfe6e285df697842540c29 is good enough for that last issue. It's just overloading typescript's inferring to help it figure out that the return type is okay here.

jakeleveroni commented:
> @jakeleveroni Sorry for the massive delay. I think the change in [4ca2c65](https://github.com/paularmstrong/onerepo/commit/4ca2c65788077dc26bdfe6e285df697842540c29) is good enough for that last issue. It's just overloading typescript's inferring to help it figure out that the return type is okay here.

No worries! hope all is well 😄 these changes look good to me if you're happy with them I saw send it!
